### PR TITLE
Simple retry on services & plugins creation.

### PR DESCRIPTION
**Pull request by cyrilgdn**
Depends on #1 

I'll try with that to see if it solves our problem but this one may be enhanced with a provider setting (like `upsert`) and  maybe retry more stuff (update, other resources, etc...). Let me know what do you think about that.

### CODE REVIEW

File: kong/resource_kong_plugin.go [hunk 1]
```diff
@@ -80,24 +81,32 @@ func resourceKongPluginCreate(d *schema.ResourceData, meta interface{}) error {
...
-				pluginClient, pluginRequest.Name, pluginRequest.ConsumerId, pluginRequest.RouteId, pluginRequest.ServiceId,
-			)
-			if err != nil {
-				return err
-			}
-
-			d.SetId(dbPlugin.Id)
-			return resourceKongPluginUpdate(d, meta)
+	var pluginID string
+	if err = retryOnce(func() error {
```
Comment (vthiery) [inline kong/resource_kong_plugin.go]:
```suggestion
	if err := retryOnce(func() error {
```

**Comment by multani** [comment]
I think we'll need this one anyway, as #1 addresses only part of our problems (subsequent Terraform runs, after the first one crashed, where Terraform didn't record it created the plugin/service).

Maybe we can retry 3 times, with [an exponential backoff + some jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) in the retry function, so we greatly reduce concurrent accesses?
We don't have thousands of them, so I think this should prevent 99.99% of our problems already.

**Comment by cyrilgdn** [comment]
@multani  I updated the PR. It now uses the [Terraform Retry helper](https://www.terraform.io/docs/extend/resources/retries-and-customizable-timeouts.html#retry) that I've just discovered (or maybe I forgot it :man_shrugging: )

It does not retry X times but it retries with a backoff until the specified timeout (no jitter though).


I also add settings to enable it in the provider, but if we want to push this upstream eventually, this means we'll probably have to retry everything ? (or to have more granular settings?)

File: kong/utils.go [hunk 2]
```diff
@@ -1,6 +1,8 @@
 package kong
 
-import "regexp"
+import (
```
Comment (vthiery) [inline kong/utils.go]: Probably don't want to do that

File: kong/resource_kong_plugin.go [hunk 3]
```diff
@@ -80,23 +82,38 @@ func resourceKongPluginCreate(d *schema.ResourceData, meta interface{}) error {
...
-	if err != nil {
-		if config.upsertResources && strings.Contains(err.Error(), "unique constraint violation") {
-			dbPlugin, err := findPlugin(
-				pluginClient, pluginRequest.Name, pluginRequest.ConsumerId, pluginRequest.RouteId, pluginRequest.ServiceId,
-			)
-			if err != nil {
-				return err
-			}
+	var pluginID string
+	if err = resource.Retry(config.retryTimeout, func() *resource.RetryError {
```
Comment (vthiery) [inline kong/resource_kong_plugin.go]:
```suggestion
	if err := resource.Retry(config.retryTimeout, func() *resource.RetryError {
```

**Comment by multani** [comment]
> but if we want to push this upstream eventually

Let's see first what will happen with how we configure Kong in the future (if we move towards the DBLess setup (which will remove a bunch of problems we have actually)), I don't think we'll use the provider anymore. Upstreaming the change will be less of a concern (we can still open an issue though.)

File: kong/resource_kong_plugin.go [hunk 4]
```diff
@@ -80,23 +82,38 @@ func resourceKongPluginCreate(d *schema.ResourceData, meta interface{}) error {
...
+		log.Printf("creating plugin %s", pluginRequest.Name)
 
-			d.SetId(dbPlugin.Id)
-			return resourceKongPluginUpdate(d, meta)
+		plugin, err := pluginClient.Create(pluginRequest)
+		if err != nil {
+			if config.upsertResources && strings.Contains(err.Error(), "unique constraint violation") {
```
Comment (vthiery) [inline kong/resource_kong_plugin.go:91]: I'd write a `isContraintViolationError` function that checks the second part but feel free to ignore.

**Review by vthiery — APPROVED** [review]

**Pull request merged**